### PR TITLE
Jonavin gmmk pro iso keymap

### PR DESCRIPTION
**Pull request by Jonavin**
## Description

ISO version of gmmk pro keymap with updated rgb_matrix_map.h for ISO layout
Thanks to RustyBrakes for the keymap adjustments and ISO Layout testing

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [ ] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by RustyBrakes** [comment]
Confirming the successful compile, flash and testing for all core functions on ISO GMMK Pro. The key and LED differences were thoroughly tested.

File: keyboards/gmmk/pro/iso/keymaps/jonavin/config.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+#ifdef RGB_MATRIX_ENABLE
+    #define RGB_MATRIX_STARTUP_MODE RGB_MATRIX_SOLID_COLOR
+    #define RGB_DISABLE_WHEN_USB_SUSPENDED
+#endif
+
+// add fifth layer for colemak  -- set "COLEMAK_LAYER_ENABLE = yes" in rules.mk to enable
+#if defined COLEMAK_LAYER_ENABLE
+    #define DYNAMIC_KEYMAP_LAYER_COUNT 5
+    #define _COLEMAK 4
+#endif // COLEMAK_LAYER_ENABLE
```
Comment (drashna) [inline keyboards/gmmk/pro/iso/keymaps/jonavin/config.h:39]:
```suggestion
#ifdef RGB_MATRIX_ENABLE
#    define RGB_MATRIX_STARTUP_MODE RGB_MATRIX_SOLID_COLOR
#    define RGB_DISABLE_WHEN_USB_SUSPENDED
#endif

// add fifth layer for colemak  -- set "COLEMAK_LAYER_ENABLE = yes" in rules.mk to enable
#if defined COLEMAK_LAYER_ENABLE
#    define DYNAMIC_KEYMAP_LAYER_COUNT 5
#    define _COLEMAK 4
#endif // COLEMAK_LAYER_ENABLE
```

File: keyboards/gmmk/pro/iso/keymaps/jonavin/keymap.c [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    ),
+
+   [_RAISE] = LAYOUT(
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,                   _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______, _______, _______,
+        _______, _______, _______,                            _______,                            _______, _______, _______, _______, _______, _______
+    ),
```
Comment (drashna) [inline keyboards/gmmk/pro/iso/keymaps/jonavin/keymap.c:66]:
```suggestion
```

File: keyboards/gmmk/pro/iso/keymaps/jonavin/keymap.c [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        }
+    }
+
+    void suspend_power_down_user(void) {
+        rgb_matrix_set_suspend_state(true);
+    }
+
+    void suspend_wakeup_init_user(void) {
+        rgb_matrix_set_suspend_state(false);
+    }
```
Comment (drashna) [inline keyboards/gmmk/pro/iso/keymaps/jonavin/keymap.c:197]:
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/gmmk/pro/iso/keymaps/jonavin/keymap.c (hunk 2 above)
Comment (Jonavin) [reply keyboards/gmmk/pro/iso/keymaps/jonavin/keymap.c:66]: Does removing the entire layer not cause problems? Or is TRANS implied if not defined?  Does it affect ability to dynamic mapping in VIA later?

**Comment by Jonavin** [comment]
Tested ok with latest commits

File: keyboards/gmmk/pro/iso/keymaps/jonavin/keymap.c [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        KC_TAB,  KC_Q,    KC_W,    KC_F,    KC_P,    KC_G,    KC_J,    KC_L,    KC_U,    KC_Y,    KC_SCLN, KC_LBRC, KC_RBRC,                   KC_PGUP,
+        TT(_LOWER), KC_A, KC_R,    KC_S,    KC_T,    KC_D,    KC_H,    KC_N,    KC_E,    KC_I,    KC_O,    KC_QUOT, KC_NUHS, KC_ENT,           KC_PGDN,
+        KC_LSFTCAPSWIN, KC_NUBS, KC_Z, KC_X, KC_C,   KC_V,    KC_B,    KC_K,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH,          KC_RSFT, KC_UP,   KC_END,
+        KC_LCTL, KC_LGUI, KC_LALT,                            KC_SPC,                             KC_RALT, MO(_FN1),KC_RCTL, KC_LEFT, KC_DOWN, KC_RGHT
+    ),
+#endif // COLEMAK_LAYER_ENABLE
+};
+
+#if defined(ENCODER_ENABLE) && !defined(ENCODER_DEFAULTACTIONS_ENABLE)     // Encoder Functionality when not using userspace defaults
+    void encoder_action_rgbhue(bool clockwise) {
```
Comment (drashna) [inline keyboards/gmmk/pro/iso/keymaps/jonavin/keymap.c:72]: This indentation bothers, me, but is more of a nitpick.

**Review by drashna — APPROVED** [review]

**Review by fauxpark — APPROVED** [review]

**Comment by BrudiBanani** [comment]
Thanks for this fantastic keymap. Not sure if this is the right place for feedback though.
For me Left ALT + Encoder does nothing. Should have next/prev track option. Tested with VLC Player and Spotify. 
Also i could not figure out how to enter colemak layer. 
Is there any possibility to debug locally?

**Comment by RustyBrakes** [comment]
hey @BrudiBanani 

I don't mean to suggest that you're wrong, but I just pulled this from scratch again, compiled/flashed, and I'm having success with both the LAlt Encoder functionality and the Colemak layer, so I'm afraid it's not a global issue. 

To access Colemak, hold left shift, then use encoder until right side LEDs are all blue (four clockwise clicks of the encoder).

It may sound obvious but did you make any changes or modify the rules.mk file?

**Comment by BrudiBanani** [comment]
@RustyBrakes Thanks for replying me. I changed nothing. Just pulled, compiled/flash as you also did. So no changes to the the rules.mk file or anything else.
Thanks for Left Shift Encoder hotkey tip :D i did not recognize that even reading the readme 5 times! (now i have found it)

Where is the encoder_action_mediatrack() function defined? Are there known problems with it regarding the OS? (I use Ubuntu)
prvs/next track from fn layer working just fine.

**Comment by Jonavin** [comment]
@BrudiBanani a lot of the code is in the userspace because I use it on many boards for consistency as I’m constantly changing keyboards. 

I’d suggest that you review the qmk documentation on this subject to answer this and any other questions you might have about customization with qmk. https://docs.qmk.fm/#/feature_userspace

**Comment by RustyBrakes** [comment]
@BrudiBanani Look under qmk_firmware/user/jonavin and you'll see the `jonavin_encoder.c` is there. The specific code relating to it is lines 37-42:

```
void encoder_action_mediatrack(bool clockwise) {
        if (clockwise)
            tap_code(KC_MEDIA_NEXT_TRACK);
        else
            tap_code(KC_MEDIA_PREV_TRACK);
    }
```

Yes, KC_MPRV and KC_MNXT are just aliases to the full keycodes like you say!


**New idea:**
Try Fn+F5 and Fn+F6 which are also mapped to KC_MPRV and KC_MNXT . This should tell you whether it's an Ubuntu problem or something about the compiling/keymap

**Comment by BrudiBanani** [comment]
@Jonavin Thanks for pointing me in the right direction. I do understand the userspace feature now. 

@RustyBrakes Thanks aswell. Fn+F5 and FN+F6 is indeed working as expected. 

I enabled debug mode now and added some dprint messages inside the if else condition to be sure that left alt + encoder combination is recognized. After compiling/reflashing i am able to see them inside qmk console.
So now I am stuck knowing Fn+F5 is working and Left Alt + Encoder is recognized correctly.
What would be my next step? :/

**Comment by RustyBrakes** [comment]
OK I've worked it out @BrudiBanani (after testing on my PopOS laptop):

Try ALT + FN + F5/F6 and you'll see that for some reason when ALT is held, the KC_MNXT and KC_MPRV don't work anymore. So you'll need to find a way to release ALT when you get to this section of code... I think I remember seeing something in the docs about that concept but I'm afraid I don't have time to go and look for it again now

**Comment by BrudiBanani** [comment]
@RustyBrakes Thank you very much! I'll have a look.

In the mean time I thought it would be nice to just swap Left Ctrl + Encoder & Left Alt + Encoder Functions.
After doing so I can successfully change to previous or next track. But more interesting is the navword function which is now cycling through my workspaces :D (actually i like that)

If I find a more permanent solution like you suggested I will add it here. 

Thanks for all your help!

**Comment by Jonavin** [comment]
> OK I've worked it out @BrudiBanani (after testing on my PopOS laptop):
> 
> Try ALT + FN + F5/F6 and you'll see that for some reason when ALT is held, the KC_MNXT and KC_MPRV don't work anymore. So you'll need to find a way to release ALT when you get to this section of code... I think I remember seeing something in the docs about that concept but I'm afraid I don't have time to go and look for it again now

That's interesting. With the ANSI keymap, I cannot duplicate this problem.  It works as intended on both Windows and iOS.

**Pull request merged**